src/SpeedJumpIrm.sol

```diff
@@ -120,43 +124,42 @@ contract AdaptativeCurveIrm is IIrm {
             int256 elapsed = int256(block.timestamp - market.lastUpdate);
             int256 linearAdaptation = speed * elapsed;
-            int256 adaptationMultiplier = MathLib.wExp(linearAdaptation);
-            // endRateAtTarget is bounded between MIN_RATE_AT_TARGET and MAX_RATE_AT_TARGET.
-            int256 endRateAtTarget =
-                startRateAtTarget.wMulDown(adaptationMultiplier).bound(MIN_RATE_AT_TARGET, MAX_RATE_AT_TARGET);
-            int256 endBorrowRate = _curve(endRateAtTarget, err);
-
-            // Then we compute the average rate over the period.
-            // Note that startBorrowRate is defined in the computations below.
-            // avgBorrowRate = 1 / elapsed * ∫ startBorrowRate * exp(speed * t) dt between 0 and elapsed
-            //               = startBorrowRate * (exp(linearAdaptation) - 1) / linearAdaptation
-            //               = (endBorrowRate - startBorrowRate) / linearAdaptation
-            // And for linearAdaptation around zero: avgBorrowRate ~ startBorrowRate = endBorrowRate.
-            // Also, when it is the first interaction (rateAtTarget = 0).
-            int256 avgBorrowRate;
-            if (linearAdaptation == 0) {
-                avgBorrowRate = endBorrowRate;
-            } else {
-                int256 startBorrowRate = _curve(startRateAtTarget, err);
-                avgBorrowRate = (endBorrowRate - startBorrowRate).wDivDown(linearAdaptation);
+
+            // We want to approximate an average of the rate over the period [0, T]:
+            // avg = 1/T ∫_0^T curve(startRateAtTarget * exp(speed * x), err) dx
+            // We approximate the integral with a Riemann sum (steps of length T/N). We want to underestimate the rate,
+            // which means doing a left Riemann (a=0, b=N-1) when the rate goes up (err>0) and a right Riemann (a=1,
+            // b=N) when the rate goes down (err<0).
+            // avg ~= 1/T Σ_i=a^b curve(startRateAtTarget * exp(speed * T/N * i), err) * T / N
+            //     ~= Σ_i=a^b curve(startRateAtTarget * exp(linearVariation/N * i), err) / N
+            // curve is linear in startRateAtTarget, so:
+            //     ~= curve(Σ_i=a^b startRateAtTarget * exp(linearVariation/N * i), err) / N
+            //     ~= curve(Σ_i=a^b startRateAtTarget * exp(linearVariation/N * i) / N, err)
+            int256 sumRateAtTarget;
+            int256 step = linearAdaptation / N_STEPS;
+            // Compute the terms 1 to N_STEPS - 1.
+            for (int256 k = 1; k < N_STEPS; k++) {
+                sumRateAtTarget += _newRateAtTarget(startRateAtTarget, step * k);
             }
+            int256 endRateAtTarget = _newRateAtTarget(startRateAtTarget, linearAdaptation);
+            // Add the term 0 of N_STEPS depending of if we do a left or right Riemann.
+            sumRateAtTarget += err < 0 ? endRateAtTarget : startRateAtTarget;
```

Comment:
```suggestion
            // Add the term 0 of N_STEPS depending on whether we do a left or right Riemann.
            if (err < 0) {
                int256 endRateAtTarget = _newRateAtTarget(startRateAtTarget, linearAdaptation);
            
                sumRateAtTarget += endRateAtTarget;
            } else {
                sumRateAtTarget += startRateAtTarget;
            }
```

Reply:
Note i also modified the comment's phrasing

Reply:
The function is returning `endRateAtTarget` so I don't think we can do this optimization

Reply:
but then you don't compute the endRateAtTarget?